diff --git a/src/circuit/commit_ivk.rs b/src/circuit/commit_ivk.rs
index 33692a4..e5cb7d3 100644
--- a/src/circuit/commit_ivk.rs
+++ b/src/circuit/commit_ivk.rs
@@ -27,7 +27,8 @@ pub struct CommitIvkChip {
 }
 
 impl CommitIvkChip {
-    pub(in crate::circuit) fn configure(
+    #[allow(missing_docs)]
+    pub fn configure(
         meta: &mut ConstraintSystem<pallas::Base>,
         advices: [Column<Advice>; 10],
     ) -> CommitIvkConfig {
@@ -222,12 +223,14 @@ impl CommitIvkChip {
         config
     }
 
-    pub(in crate::circuit) fn construct(config: CommitIvkConfig) -> Self {
+    #[allow(missing_docs)]
+    pub fn construct(config: CommitIvkConfig) -> Self {
         Self { config }
     }
 }
 
-pub(in crate::circuit) mod gadgets {
+#[allow(missing_docs)]
+pub mod gadgets {
     use halo2_gadgets::utilities::{lookup_range_check::LookupRangeCheckConfig, RangeConstrained};
     use halo2_proofs::circuit::Chip;
 
@@ -238,7 +241,7 @@ pub(in crate::circuit) mod gadgets {
     /// [Section 5.4.8.4 Sinsemilla commitments]: https://zips.z.cash/protocol/protocol.pdf#concretesinsemillacommit
     #[allow(non_snake_case)]
     #[allow(clippy::type_complexity)]
-    pub(in crate::circuit) fn commit_ivk(
+    pub fn commit_ivk(
         sinsemilla_chip: SinsemillaChip<
             OrchardHashDomains,
             OrchardCommitDomains,
diff --git a/src/circuit/gadget/add_chip.rs b/src/circuit/gadget/add_chip.rs
index 0973a3f..e1f4e26 100644
--- a/src/circuit/gadget/add_chip.rs
+++ b/src/circuit/gadget/add_chip.rs
@@ -7,8 +7,9 @@ use pasta_curves::pallas;
 
 use super::AddInstruction;
 
+#[allow(missing_docs)]
 #[derive(Clone, Debug)]
-pub(in crate::circuit) struct AddConfig {
+pub struct AddConfig {
     a: Column<Advice>,
     b: Column<Advice>,
     c: Column<Advice>,
@@ -16,7 +17,8 @@ pub(in crate::circuit) struct AddConfig {
 }
 
 /// A chip implementing a single addition constraint `c = a + b` on a single row.
-pub(in crate::circuit) struct AddChip {
+#[derive(Clone, Debug)]
+pub struct AddChip {
     config: AddConfig,
 }
 
@@ -34,7 +36,7 @@ impl Chip<pallas::Base> for AddChip {
 }
 
 impl AddChip {
-    pub(in crate::circuit) fn configure(
+    pub fn configure(
         meta: &mut ConstraintSystem<pallas::Base>,
         a: Column<Advice>,
         b: Column<Advice>,
@@ -53,7 +55,7 @@ impl AddChip {
         AddConfig { a, b, c, q_add }
     }
 
-    pub(in crate::circuit) fn construct(config: AddConfig) -> Self {
+    pub fn construct(config: AddConfig) -> Self {
         Self { config }
     }
 }
diff --git a/src/circuit/gadget.rs b/src/circuit/gadget.rs
index 12c3c90..d851bdc 100644
--- a/src/circuit/gadget.rs
+++ b/src/circuit/gadget.rs
@@ -24,7 +24,8 @@ use halo2_proofs::{
     plonk::{self, Advice, Assigned, Column},
 };
 
-pub(in crate::circuit) mod add_chip;
+#[allow(missing_docs)]
+pub mod add_chip;
 
 impl super::Config {
     pub(super) fn add_chip(&self) -> add_chip::AddChip {
@@ -77,7 +78,7 @@ impl super::Config {
 }
 
 /// An instruction set for adding two circuit words (field elements).
-pub(in crate::circuit) trait AddInstruction<F: Field>: Chip<F> {
+pub trait AddInstruction<F: Field>: Chip<F> {
     /// Constraints `a + b` and returns the sum.
     fn add(
         &self,
@@ -92,7 +93,7 @@ pub(in crate::circuit) trait AddInstruction<F: Field>: Chip<F> {
 /// Usages of this helper are technically superfluous, as the single-cell region is only
 /// ever used in equality constraints. We could eliminate them with a
 /// [write-on-copy abstraction](https://github.com/zcash/halo2/issues/334).
-pub(in crate::circuit) fn assign_free_advice<F: Field, V: Copy>(
+pub fn assign_free_advice<F: Field, V: Copy>(
     mut layouter: impl Layouter<F>,
     column: Column<Advice>,
     value: Value<V>,
@@ -109,7 +110,7 @@ where
 /// `ValueCommit^Orchard` from [Section 5.4.8.3 Homomorphic Pedersen commitments (Sapling and Orchard)].
 ///
 /// [Section 5.4.8.3 Homomorphic Pedersen commitments (Sapling and Orchard)]: https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
-pub(in crate::circuit) fn value_commit_orchard<
+pub fn value_commit_orchard<
     EccChip: EccInstructions<
         pallas::Affine,
         FixedPoints = OrchardFixedBases,
@@ -145,7 +146,7 @@ pub(in crate::circuit) fn value_commit_orchard<
 ///
 /// [Section 4.16: Note Commitments and Nullifiers]: https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers
 #[allow(clippy::too_many_arguments)]
-pub(in crate::circuit) fn derive_nullifier<
+pub fn derive_nullifier<
     PoseidonChip: PoseidonSpongeInstructions<pallas::Base, poseidon::P128Pow5T3, ConstantLength<2>, 3, 2>,
     AddChip: AddInstruction<pallas::Base>,
     EccChip: EccInstructions<
@@ -199,5 +200,7 @@ pub(in crate::circuit) fn derive_nullifier<
         .map(|res| res.extract_p())
 }
 
-pub(in crate::circuit) use crate::circuit::commit_ivk::gadgets::commit_ivk;
-pub(in crate::circuit) use crate::circuit::note_commit::gadgets::note_commit;
+#[allow(missing_docs)]
+pub use crate::circuit::commit_ivk::gadgets::commit_ivk;
+#[allow(missing_docs)]
+pub use crate::circuit::note_commit::gadgets::note_commit;
diff --git a/src/circuit/note_commit.rs b/src/circuit/note_commit.rs
index 9daa685..9571030 100644
--- a/src/circuit/note_commit.rs
+++ b/src/circuit/note_commit.rs
@@ -1437,7 +1437,8 @@ pub struct NoteCommitChip {
 impl NoteCommitChip {
     #[allow(non_snake_case)]
     #[allow(clippy::many_single_char_names)]
-    pub(in crate::circuit) fn configure(
+    #[allow(missing_docs)]
+    pub fn configure(
         meta: &mut ConstraintSystem<pallas::Base>,
         advices: [Column<Advice>; 10],
         sinsemilla_config: SinsemillaConfig<
@@ -1556,12 +1557,14 @@ impl NoteCommitChip {
         }
     }
 
-    pub(in crate::circuit) fn construct(config: NoteCommitConfig) -> Self {
+    #[allow(missing_docs)]
+    pub fn construct(config: NoteCommitConfig) -> Self {
         Self { config }
     }
 }
 
-pub(in crate::circuit) mod gadgets {
+#[allow(missing_docs)]
+pub mod gadgets {
     use halo2_proofs::circuit::{Chip, Value};
 
     use super::*;
@@ -1569,7 +1572,7 @@ pub(in crate::circuit) mod gadgets {
     #[allow(clippy::many_single_char_names)]
     #[allow(clippy::type_complexity)]
     #[allow(clippy::too_many_arguments)]
-    pub(in crate::circuit) fn note_commit(
+    pub fn note_commit(
         mut layouter: impl Layouter<pallas::Base>,
         chip: SinsemillaChip<OrchardHashDomains, OrchardCommitDomains, OrchardFixedBases>,
         ecc_chip: EccChip<OrchardFixedBases>,
diff --git a/src/circuit.rs b/src/circuit.rs
index bdfead6..807977f 100644
--- a/src/circuit.rs
+++ b/src/circuit.rs
@@ -58,9 +58,11 @@ use halo2_gadgets::{
     utilities::lookup_range_check::LookupRangeCheckConfig,
 };
 
-mod commit_ivk;
+#[allow(missing_docs)]
+pub mod commit_ivk;
 pub mod gadget;
-mod note_commit;
+#[allow(missing_docs)]
+pub mod note_commit;
 
 pub use crate::Proof;
 
diff --git a/src/lib.rs b/src/lib.rs
index 3a52af5..d96f226 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -48,6 +48,15 @@ mod test_vectors;
 pub use action::Action;
 pub use address::Address;
 pub use bundle::Bundle;
+/// Orchard fixed-base selector used by circuit scalar multiplication gadgets.
+pub type OrchardFixedBases = constants::fixed_bases::OrchardFixedBases;
+/// Orchard full-width fixed bases used by circuit scalar multiplication gadgets.
+pub type OrchardFixedBasesFull = constants::fixed_bases::OrchardFixedBasesFull;
+pub use constants::fixed_bases::{NullifierK, ValueCommitV};
+/// Orchard Sinsemilla hash domains.
+pub type OrchardHashDomains = constants::sinsemilla::OrchardHashDomains;
+/// Orchard Sinsemilla commit domains.
+pub type OrchardCommitDomains = constants::sinsemilla::OrchardCommitDomains;
 pub use constants::MERKLE_DEPTH_ORCHARD as NOTE_COMMITMENT_TREE_DEPTH;
 pub use note::Note;
 pub use tree::Anchor;
diff --git a/src/note/nullifier.rs b/src/note/nullifier.rs
index 15ac369..a1658cc 100644
--- a/src/note/nullifier.rs
+++ b/src/note/nullifier.rs
@@ -54,8 +54,10 @@ impl Nullifier {
         rho: pallas::Base,
         psi: pallas::Base,
         cm: NoteCommitment,
+        domain_prefix: &str,
+        value: &[u8]
     ) -> Self {
-        let k = pallas::Point::hash_to_curve("z.cash:Orchard")(b"K");
+        let k = pallas::Point::hash_to_curve(domain_prefix)(value);
 
         Nullifier(extract_p(&(k * mod_r_p(nk.prf_nf(rho) + psi) + cm.0)))
     }
diff --git a/src/note.rs b/src/note.rs
index 6a9c70b..2a09af2 100644
--- a/src/note.rs
+++ b/src/note.rs
@@ -54,7 +54,8 @@ impl Rho {
         Rho(nf.0)
     }
 
-    pub(crate) fn into_inner(self) -> pallas::Base {
+    /// Returns the underlying field element.
+    pub fn into_inner(self) -> pallas::Base {
         self.0
     }
 }
@@ -92,7 +93,7 @@ impl RandomSeed {
     /// Defined in [Zcash Protocol Spec § 4.7.3: Sending Notes (Orchard)][orchardsend].
     ///
     /// [orchardsend]: https://zips.z.cash/protocol/nu5.pdf#orchardsend
-    pub(crate) fn psi(&self, rho: &Rho) -> pallas::Base {
+    pub fn psi(&self, rho: &Rho) -> pallas::Base {
         to_base(PrfExpand::PSI.with(&self.0, &rho.to_bytes()))
     }
 
@@ -121,6 +122,11 @@ impl RandomSeed {
             PrfExpand::ORCHARD_RCM.with(&self.0, &rho.to_bytes()),
         ))
     }
+
+    /// Returns the scalar used as note commitment trapdoor for this seed and rho.
+    pub fn rcm_scalar(&self, rho: &Rho) -> pallas::Scalar {
+        self.rcm(rho).inner()
+    }
 }
 
 /// A discrete amount of funds received by an address.
@@ -287,6 +293,21 @@ impl Note {
             self.rho.0,
             self.rseed.psi(&self.rho),
             self.commitment(),
+            "z.cash:Orchard",
+            b"K",
+        )
+    }
+
+    /// A nullifier that will be derived from the note, but it will never
+    /// appear in a transaction.
+    pub fn hiding_nullifier(&self, fvk: &FullViewingKey, domain: &str, value: &[u8]) -> Nullifier {
+        Nullifier::derive(
+            fvk.nk(),
+            self.rho.0,
+            self.rseed.psi(&self.rho),
+            self.commitment(),
+            domain,
+            value,
         )
     }
 }
diff --git a/src/value.rs b/src/value.rs
index 516bf8e..91f1031 100644
--- a/src/value.rs
+++ b/src/value.rs
@@ -239,7 +239,8 @@ impl TryFrom<ValueSum> for i64 {
 pub struct ValueCommitTrapdoor(pallas::Scalar);
 
 impl ValueCommitTrapdoor {
-    pub(crate) fn inner(&self) -> pallas::Scalar {
+    /// Returns the underlying scalar value.
+    pub fn inner(&self) -> pallas::Scalar {
         self.0
     }
 
@@ -375,7 +376,7 @@ impl ValueCommitment {
     }
 
     /// x-coordinate of this value commitment.
-    pub(crate) fn x(&self) -> pallas::Base {
+    pub fn x(&self) -> pallas::Base {
         if self.0 == pallas::Point::identity() {
             pallas::Base::zero()
         } else {
@@ -384,7 +385,7 @@ impl ValueCommitment {
     }
 
     /// y-coordinate of this value commitment.
-    pub(crate) fn y(&self) -> pallas::Base {
+    pub fn y(&self) -> pallas::Base {
         if self.0 == pallas::Point::identity() {
             pallas::Base::zero()
         } else {
