diff --git i/src/circuit.rs w/src/circuit.rs
index d2f083e..c09d93b 100644
--- i/src/circuit.rs
+++ w/src/circuit.rs
@@ -25,9 +25,9 @@ use crate::{value::NoteValue, PaymentAddress, ProofGenerationKey};
 #[cfg(test)]
 use group::ff::PrimeFieldBits;
 
-mod constants;
-mod ecc;
-mod pedersen_hash;
+pub mod constants;
+pub mod ecc;
+pub mod pedersen_hash;
 
 /// The opening (value and randomness) of a Sapling value commitment.
 #[derive(Clone)]
@@ -592,7 +592,7 @@ impl SpendVerifyingKey {
 
 /// The verifying key for the Sapling Spend circuit, with precomputations optimized for
 /// verifying individual proofs.
-pub struct PreparedSpendVerifyingKey(pub(crate) groth16::PreparedVerifyingKey<Bls12>);
+pub struct PreparedSpendVerifyingKey(pub groth16::PreparedVerifyingKey<Bls12>);
 
 /// The parameters for the Sapling Output circuit.
 pub struct OutputParameters(pub(crate) groth16::Parameters<Bls12>);
diff --git i/src/keys.rs w/src/keys.rs
index 4c357f8..723a751 100644
--- i/src/keys.rs
+++ w/src/keys.rs
@@ -170,7 +170,7 @@ impl SpendValidatingKey {
     }
 
     /// Parses a `SpendValidatingKey` from its encoded form.
-    pub(crate) fn from_bytes(bytes: &[u8]) -> Option<Self> {
+    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
         <[u8; 32]>::try_from(bytes)
             .ok()
             .and_then(|b| {
diff --git i/src/note.rs w/src/note.rs
index 70d1473..7699d59 100644
--- i/src/note.rs
+++ w/src/note.rs
@@ -113,6 +113,15 @@ impl Note {
         Nullifier::derive(nk, self.cm_full_point(), position)
     }
 
+    pub fn nf_hiding(
+        &self,
+        nk: &NullifierDerivingKey,
+        position: u64,
+        personalization: &[u8],
+    ) -> Nullifier {
+        Nullifier::derive_hiding(nk, self.cm_full_point(), position, personalization)
+    }
+
     /// Computes the note commitment
     pub fn cmu(&self) -> ExtractedNoteCommitment {
         self.cm_full_point().into()
diff --git i/src/note/nullifier.rs w/src/note/nullifier.rs
index 58fd01b..f96cef2 100644
--- i/src/note/nullifier.rs
+++ w/src/note/nullifier.rs
@@ -7,7 +7,7 @@ use subtle::{Choice, ConstantTimeEq};
 use super::NoteCommitment;
 use crate::{
     keys::NullifierDerivingKey,
-    spec::{mixing_pedersen_hash, prf_nf},
+    spec::{mixing_pedersen_hash, prf_nf, prf_nf_hiding},
 };
 
 /// Typesafe wrapper for nullifier values.
@@ -40,6 +40,17 @@ impl Nullifier {
         let rho = mixing_pedersen_hash(cm.inner(), position);
         Nullifier(prf_nf(&nk.0, &rho))
     }
+
+    // TODO: add docs
+    pub(super) fn derive_hiding(
+        nk: &NullifierDerivingKey,
+        cm: NoteCommitment,
+        position: u64,
+        personalization: &[u8],
+    ) -> Self {
+        let rho = mixing_pedersen_hash(cm.inner(), position);
+        Nullifier(prf_nf_hiding(&nk.0, &rho, personalization))
+    }
 }
 
 impl AsRef<[u8]> for Nullifier {
diff --git i/src/spec.rs w/src/spec.rs
index eabf17a..99ff424 100644
--- i/src/spec.rs
+++ w/src/spec.rs
@@ -65,9 +65,18 @@ pub(crate) fn mixing_pedersen_hash(
 ///
 /// [concreteprfs]: https://zips.z.cash/protocol/protocol.pdf#concreteprfs
 pub(crate) fn prf_nf(nk: &jubjub::SubgroupPoint, rho: &jubjub::SubgroupPoint) -> [u8; 32] {
+    prf_nf_hiding(nk, rho, PRF_NF_PERSONALIZATION)
+}
+
+/// prf_nf with custom personalization
+pub(crate) fn prf_nf_hiding(
+    nk: &jubjub::SubgroupPoint,
+    rho: &jubjub::SubgroupPoint,
+    prf_nf_personalization: &[u8],
+) -> [u8; 32] {
     Blake2sParams::new()
         .hash_length(32)
-        .personal(PRF_NF_PERSONALIZATION)
+        .personal(prf_nf_personalization)
         .to_state()
         .update(&nk.to_bytes())
         .update(&rho.to_bytes())
